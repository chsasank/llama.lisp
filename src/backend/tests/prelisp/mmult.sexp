(c-lisp
    ;; Function to add dot products for a 4x4 block
    (define ((__add_dot4x4 void)
            (k int)
            (a (ptr float))
            (lda int)
            (b (ptr float))
            (ldb int)
            (c (ptr float))
            (ldc int))

        ,@(init_zeros c ldc)

        (for ((set p 0) (lt p k) (set p (add p 1)))
            ,@(store_vals p a b c lda ldb ldc))

        (ret))
        
    
    (define ((__kernel void)
            (a (ptr float))
            (b (ptr float))
            (c (ptr float))
            (m int)
            (n int)
            (k int))

        (declare i int)
        (declare j int)

        (declare lda int)
        (declare ldb int)
        (declare ldc int)

        (set lda m)
        (set ldb n)
        (set ldc k)

        (for ((set j 0) (lt j n) (set j (add j 4)))
            (for ((set i 0) (lt i m) (set i (add i 4)))
                (call __add_dot4x4 k 
                    (ptradd a (add i (mul 0 lda))) lda 
                    (ptradd b (mul j ldb)) ldb
                    (ptradd c (add i (mul j ldc))) ldc)))
        (ret)))
