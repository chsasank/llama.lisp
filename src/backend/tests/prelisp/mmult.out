(c-lisp
  (define ((__add_dot4x4 void)
           (k int)
           (a (ptr float))
           (lda int)
           (b (ptr float))
           (ldb int)
           (c (ptr float))
           (ldc int))
    (store (ptradd c (add 0 (mul 0 ldc))) 0)
    (store (ptradd c (add 0 (mul 1 ldc))) 0)
    (store (ptradd c (add 0 (mul 2 ldc))) 0)
    (store (ptradd c (add 0 (mul 3 ldc))) 0)
    (store (ptradd c (add 1 (mul 0 ldc))) 0)
    (store (ptradd c (add 1 (mul 1 ldc))) 0)
    (store (ptradd c (add 1 (mul 2 ldc))) 0)
    (store (ptradd c (add 1 (mul 3 ldc))) 0)
    (store (ptradd c (add 2 (mul 0 ldc))) 0)
    (store (ptradd c (add 2 (mul 1 ldc))) 0)
    (store (ptradd c (add 2 (mul 2 ldc))) 0)
    (store (ptradd c (add 2 (mul 3 ldc))) 0)
    (store (ptradd c (add 3 (mul 0 ldc))) 0)
    (store (ptradd c (add 3 (mul 1 ldc))) 0)
    (store (ptradd c (add 3 (mul 2 ldc))) 0)
    (store (ptradd c (add 3 (mul 3 ldc))) 0)
    (for ((set p 0) (lt p k) (set p (add p 1)))
         (store (ptradd c (add 0 (mul 0 ldc)))
                (fadd (load (ptradd c (add 0 (mul 0 ldc))))
                      (fmul (load (ptradd a (add 0 (mul p lda))))
                            (load (ptradd b (add p (mul 0 ldb)))))))
         (store (ptradd c (add 0 (mul 1 ldc)))
                (fadd (load (ptradd c (add 0 (mul 1 ldc))))
                      (fmul (load (ptradd a (add 0 (mul p lda))))
                            (load (ptradd b (add p (mul 1 ldb)))))))
         (store (ptradd c (add 0 (mul 2 ldc)))
                (fadd (load (ptradd c (add 0 (mul 2 ldc))))
                      (fmul (load (ptradd a (add 0 (mul p lda))))
                            (load (ptradd b (add p (mul 2 ldb)))))))
         (store (ptradd c (add 0 (mul 3 ldc)))
                (fadd (load (ptradd c (add 0 (mul 3 ldc))))
                      (fmul (load (ptradd a (add 0 (mul p lda))))
                            (load (ptradd b (add p (mul 3 ldb)))))))
         (store (ptradd c (add 1 (mul 0 ldc)))
                (fadd (load (ptradd c (add 1 (mul 0 ldc))))
                      (fmul (load (ptradd a (add 1 (mul p lda))))
                            (load (ptradd b (add p (mul 0 ldb)))))))
         (store (ptradd c (add 1 (mul 1 ldc)))
                (fadd (load (ptradd c (add 1 (mul 1 ldc))))
                      (fmul (load (ptradd a (add 1 (mul p lda))))
                            (load (ptradd b (add p (mul 1 ldb)))))))
         (store (ptradd c (add 1 (mul 2 ldc)))
                (fadd (load (ptradd c (add 1 (mul 2 ldc))))
                      (fmul (load (ptradd a (add 1 (mul p lda))))
                            (load (ptradd b (add p (mul 2 ldb)))))))
         (store (ptradd c (add 1 (mul 3 ldc)))
                (fadd (load (ptradd c (add 1 (mul 3 ldc))))
                      (fmul (load (ptradd a (add 1 (mul p lda))))
                            (load (ptradd b (add p (mul 3 ldb)))))))
         (store (ptradd c (add 2 (mul 0 ldc)))
                (fadd (load (ptradd c (add 2 (mul 0 ldc))))
                      (fmul (load (ptradd a (add 2 (mul p lda))))
                            (load (ptradd b (add p (mul 0 ldb)))))))
         (store (ptradd c (add 2 (mul 1 ldc)))
                (fadd (load (ptradd c (add 2 (mul 1 ldc))))
                      (fmul (load (ptradd a (add 2 (mul p lda))))
                            (load (ptradd b (add p (mul 1 ldb)))))))
         (store (ptradd c (add 2 (mul 2 ldc)))
                (fadd (load (ptradd c (add 2 (mul 2 ldc))))
                      (fmul (load (ptradd a (add 2 (mul p lda))))
                            (load (ptradd b (add p (mul 2 ldb)))))))
         (store (ptradd c (add 2 (mul 3 ldc)))
                (fadd (load (ptradd c (add 2 (mul 3 ldc))))
                      (fmul (load (ptradd a (add 2 (mul p lda))))
                            (load (ptradd b (add p (mul 3 ldb)))))))
         (store (ptradd c (add 3 (mul 0 ldc)))
                (fadd (load (ptradd c (add 3 (mul 0 ldc))))
                      (fmul (load (ptradd a (add 3 (mul p lda))))
                            (load (ptradd b (add p (mul 0 ldb)))))))
         (store (ptradd c (add 3 (mul 1 ldc)))
                (fadd (load (ptradd c (add 3 (mul 1 ldc))))
                      (fmul (load (ptradd a (add 3 (mul p lda))))
                            (load (ptradd b (add p (mul 1 ldb)))))))
         (store (ptradd c (add 3 (mul 2 ldc)))
                (fadd (load (ptradd c (add 3 (mul 2 ldc))))
                      (fmul (load (ptradd a (add 3 (mul p lda))))
                            (load (ptradd b (add p (mul 2 ldb)))))))
         (store (ptradd c (add 3 (mul 3 ldc)))
                (fadd (load (ptradd c (add 3 (mul 3 ldc))))
                      (fmul (load (ptradd a (add 3 (mul p lda))))
                            (load (ptradd b (add p (mul 3 ldb))))))))
    (ret))
  (define ((__kernel void)
           (a (ptr float))
           (b (ptr float))
           (c (ptr float))
           (m int)
           (n int)
           (k int))
    (declare i int)
    (declare j int)
    (declare lda int)
    (declare ldb int)
    (declare ldc int)
    (set lda m)
    (set ldb n)
    (set ldc k)
    (for ((set j 0) (lt j n) (set j (add j 4)))
         (for ((set i 0) (lt i m) (set i (add i 4)))
              (call __add_dot4x4
                    k
                    (ptradd a (add i (mul 0 lda)))
                    lda
                    (ptradd b (mul j ldb))
                    ldb
                    (ptradd c (add i (mul j ldc)))
                    ldc)))
    (ret)))
